urlfc1c29991b_nil_mish_InJava_tgz-pJ8-edu_usc_cs_nsl_indexing_test_TestTrieProtoJ8/src/edu/usc/cs/nsl/counting/CountingUtil.java

```diff
@@ -0,0 +1,494 @@
+					if(the_word.equals("0017d7d14cbaf2c1c94de80ef039")){
+						the_flag = true;
+						System.out.println("Indexes "+big_index+"\t"+this_index);
+					}
+					if(count_array[this_index] == 0){ // new word
+						word_array[this_index] = the_word;
+						count_array[this_index] = 1;
+						if(the_flag)
+							System.out.println("in 0");
+					}else{
+						if(word_array[this_index].equals(the_word)){ //word seen increment
+							if(count_array[this_index] > 0)
+								count_array[this_index] += 1;
+							else if(count_array[this_index] != Long.MIN_VALUE){
+								count_array[this_index] -= 1;
+							}
+							if(the_flag)
+								System.out.println("in equals");
+						} else{ // collision (True collision)
+							if(the_flag)
+								System.out.println("in collision");
+							int the_other_big_index = (int)Math.abs(getHash(word_array[this_index])%big_array_size);
+							if(the_other_big_index == big_index){
+								count_array[this_index] = Long.MIN_VALUE;
+								num_collision += 2;
+								bw_collision.write(big_index+"\n");
+								bw_collision.flush();
+							}else {
+								if(count_array[this_index] > 0){
+									if(the_flag)
+										System.out.println("in collision: initial");
+									count_array[this_index] = -1*count_array[this_index];
+									num_collision += 2;
+									bw_collision.write(big_index+"\n");
+									bw_collision.flush();
+								}
+								else{
+									if(the_flag)
+										System.out.println("in collision: additional "+big_index+"\t"+this_index);
+									bw_collision.write(big_index+"\n");
+									bw_collision.flush();
+									num_collision++;
+									//System.out.println(the_word+" Do I come here? "+count_array[this_index]);
+								}
+							}
+						}
+					}
+				}
+				the_line = br.readLine();
+			}
+			int written_this_chunk = 0;
+			BufferedWriter bw_output = new BufferedWriter(new FileWriter(output_file,true));
+			for(int i=0; i < chunk_size; i++){
+				if(count_array[i] > 0){
+					bw_output.write(word_array[i]+"\t"+count_array[i]+"\n");
+					written_this_chunk++;
+				}else if(count_array[i] < 0){
+					if(count_array[i] != Long.MIN_VALUE){
+						bw_output.write(word_array[i]+"\t"+(-1*count_array[i])+"\n");
+						written_this_chunk++;
+					}
+				}
+			}
+			System.out.println("Processed this chunk: "+processed_this_chunk+"\tWritten this chunk: "+written_this_chunk+"\t Collisions: "+num_collision);
+			bw_output.close();
+		}
+		bw_collision.close();
+		processCollision(max_string_size,output_file,num_collision,0,num_collision);
+		//processCollisionSimple(max_string_size,output_file,num_collision);
+	}
+	
+	/**
+	 * Simple counting function (not used)
+	 * @param max_string_size
+	 * @param output_file
+	 * @param num_collisions
+	 * @throws IOException
+	 */
+	private void processCollisionSimple(int max_string_size,String output_file,int num_collisions) throws IOException{
+		BufferedWriter bw_output = new BufferedWriter(new FileWriter(output_file,true));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *bw\_output*. The resource is referenced at lines 371, 406. The resource is closed at line 409. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *bw\_output* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).